working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+
+``` dart
+class C1<T> implements List<T> {
+  final List<T> v;
+}
+
+class C2<T> implements List<T> {
+  final List<Object> v;
+}
+
+C1<Shareable>(); // ok
+C2<Shareable>(); // runtime error: static type of C2.v is not shareable
+```
+
+A simpler alternative would be to introduce a set of shareable collection types: `ShareableList<T>`, `ShareableSet<T>`,  `ShareableMap<K, V>`, all with appropriate type parameter bounds restricting elements to subtypes of `Shareable?`. Builtin implementation of `List<T>` will also implement `ShareableList<T>` when `T` is shareable, but there would be no implicit expectation of shareability built into `List<T>` itself.
+
+### Shareable core types
+
+The following types will become shareable:
+
+* `num` (`int` and `double`), `String`, `bool`, `Null`
+* `Enum` - meaning that all enums will be shareable, this is fine because enums are deeply immutable. 
+* `RegExp`, `DateTime`, `Uri` - these might require work to be shareable, but it makes sense to allow sharing them.
+* `TypedData` - which makes all typed data types shareable.
+  * **BREAKING CHANGE**: making `TypedData` shareable changes the behaviour of `SendPort` which will stop copying it. It's unclear if we want to maintain old behaviour for compatibility reasons. One option here is to say that `TypedData` is only shared when sending through `SendPort` iff it is a member of another explicitly `Shareable` type.    
+* `Pointer` 
+
+## Shared Isolates
+
+Lets take another look at the following example: 
+
+```dart
+int global = 0; 
+
+void main() async {
+  global = 42;
+  await Isolate.run(() {
+    print(global);  // => 0
+    global = 24;
+  });  
+  print(global);  // => 42
+}
+```
+
+Stripped to the bare minimum the example does not seem to behave in a confusing way: it seems obvious that each isolate has its own version of `global` variable and mutations of `global` are not visible across isolates. However, in the real world code such behavior might be hidden deep inside a third party dependency and thus much harder to detect and understand. This behavior also makes interoperability with native code more awkward than it ought to be: calling Dart requires an isolate, something that native code does not really know or care about. 
+
+```dart
+int global;
+
+@pragma('vm:entry-point')
+void foo() => global++;
+```
+
+`@shared` global variables allow developers to tackle this problem - but hidden dependency on global state might introduce hard to diagnose and debug bugs. 
+
+We propose to tackle this problem by introducing the concept of _shared isolate_: **code running in a _shared isolate_ can only access `@shared` state and not any of isolated state, an attempt to access isolated state results in a dynamic `IsolationError`**.
```

Comment:
Shared isolate might be many isolates - it is simply not observable to the user because they automatically share `shared` variables and can't access non-shared variables.